### PR TITLE
Significantly reduce runtime of fail_compilation tests & ~double CircleCI performance

### PR DESCRIPTION
**Pull request by kinke**
`fail_compilation/fail12485.d` took about 45 seconds on my Windows box. Redirecting stderr to nirvana gets it down to about 10 seconds. The overall `start_fail_compilation_tests` runtime (with `-j4`) was reduced by about 25 seconds.

[The reasons are 2 huge lines in the 3 lines of error output, each ~5.5M chars long.]

### CODE REVIEW

File: test/tools/d_do_test.d [hunk 1]
```diff
@@ -16,6 +16,8 @@ import std.stdio;
 import std.string;
 import core.sys.posix.sys.wait;
 
+//version = MeasureTime;
```
Comment (wilzbach) [inline test/tools/d_do_test.d]: Maybe use an environment variable, s.t. this is easier to enable and the code doesn't rot (i.e. is at least checked by the compiler).

File: test/Makefile [hunk 2]
```diff
@@ -133,7 +133,7 @@ runnable_test_results=$(addsuffix .out,$(addprefix $(RESULTS_DIR)/,$(runnable_te
 compilable_tests=$(wildcard compilable/*.d) $(wildcard compilable/*.sh)
 compilable_test_results=$(addsuffix .out,$(addprefix $(RESULTS_DIR)/,$(compilable_tests)))
 
-fail_compilation_tests=$(wildcard fail_compilation/*.d) $(wildcard fail_compilation/*.html)
+fail_compilation_tests=$(wildcard fail_compilation/*.d) $(wildcard fail_compilation/*.sh) $(wildcard fail_compilation/*.html)
```
Comment (wilzbach) [inline test/Makefile]: Looks like there were other untested shell scripts in fail_compilation :/

Comment (kinke) [reply test/Makefile]: Yep, one more, added 4 weeks ago (converted from .d): https://github.com/dlang/dmd/pull/9231/files#diff-c79d6f5eac3478fdb49375949df1ffa4

**Comment by kinke** [comment]
Changed to an env variable - use `PRINT_RUNTIME=1` to produce something like:
```
 ... runnable\test12.d              -g -link-defaultlib-debug (-unittest -O -release -inline)   [28.252 secs]
 ... runnable\test9259.d            -g -link-defaultlib-debug (-inline -release -O -d -dw -de)   [48.315 secs]
 ... runnable\test17338.d           -g -link-defaultlib-debug ()   [20.058 secs]
 ... runnable\xtest46.d             -g -link-defaultlib-debug (-unittest -O -release -inline)   [54.910 secs]
 ... runnable\sdtor.d               -transition=dtorfields -g -link-defaultlib-debug (-unittest -O -release -inline)   [64.570 secs]
 ... runnable\link2644.d            -g -link-defaultlib-debug (-version=X -inline -release -O)   [16.701 secs]
```

I also added support for a `DMD_TESTSUITE_MAKE_ARGS` env variable, something I've been using for a long time for LDC, to parallelize a single test run (LDC runs the DMD testsuite up to 4 times, 32/64-bit debug/release, but serially in order to end up with a sane output). I added some further parallelization tweaks I plan on using for LDC (the AArch64 CI box features 96 slow cores...).

**Comment by kinke** [comment]
Btw, a major obstacle for good CPU scaling of these tests are the `PERMUTE_ARGS`. Each test file is processed in isolation, i.e., the 2^6=64 permutations for `runnable/xtest46.d` are run serially (by `d_do_test`, not managed by make).

File: test/Makefile (hunk 2 above)
Comment (kinke) [reply test/Makefile]: Fixed it by trimming the trailing `\r` on Windows.

**Review by WalterBright — APPROVED** [review]
This PR seems to do 3 independent things. I would have preferred they be 3 PRs.

**Comment by kinke** [comment]
Yeah I know, just my bloodhound way of doing things. ;)

Circle speed-up, times vary wildly (so 2 runtimes per config to give some idea):
* pic: {11:36, 11:49} -> {4:55, 6:30}
* no_pic: {15:21, 15:57} -> {7:49, 10:27}

Is there possibly similar low-hanging fruit for the auto-tester?

**Comment by thewilsonator** [comment]
This good to go?

**Comment by kinke** [comment]
[I'm done with this. I haven't found the autotester script, so I'm just going to assume it's not in the repo.]

There's a bit of a catch with this, as the first output can now take quite a while (something like 20 secs), as the long-running tests are started first. `runnable/xtest46.d` with its many permutations takes about 8 minutes on the AArch64 CI box, defeating all gains by having 96 cores.

**Comment by jacob-carlborg** [comment]
> I'm done with this. I haven't found the autotester script, so I'm just going to assume it's not in the repo.

Here's what the autotester is running: https://github.com/braddr/at-client/tree/master/src.

**Comment by look-at-me** [comment]
@kinke Where do you have a 96 core CI box?

**Comment by kinke** [comment]
https://app.shippable.com/github/ldc-developers/ldc/dashboard

What's really impressive are the [LLVM builds](https://app.shippable.com/github/ldc-developers/llvm/dashboard), compiling in under 11 minutes with 64 parallel jobs (as the available RAM is either 32 or 128 GB, I had to reduce their number a bit in case we hit a box with 32).

File: test/Makefile [hunk 3]
```diff
@@ -164,23 +164,30 @@ unit_tests: $(RESULTS_DIR)/unit_test_runner$(EXE)
...
 
 run_compilable_tests: $(compilable_test_results)
 
 start_compilable_tests: $(RESULTS_DIR)/.created $(test_tools)
 	@echo "Running compilable tests"
-	$(QUIET)$(MAKE) --no-print-directory run_compilable_tests
+	$(QUIET)$(MAKE) $(DMD_TESTSUITE_MAKE_ARGS) --no-print-directory run_compilable_tests
 
-run_fail_compilation_tests: $(fail_compilation_test_results)
+run_fail_compilation_tests: $(RESULTS_DIR)/fail_compilation/fail12485.sh.out $(fail_compilation_test_results)
```
Comment (wilzbach) [inline test/Makefile]: Why is this change required? (i.e. why `.out` and not `.sh`)

Comment (kinke) [reply test/Makefile]: This just makes sure `fail12485.sh` is started first in the `fail_compilation` tests block, as it takes about as much time as all other fail tests combined (& parallelized).

Comment (wilzbach) [reply test/Makefile]: Can we add this as a comment so that people in the future aren't confused? Thanks!

Comment (kinke) [reply test/Makefile]: It's 13 lines up, I assumed it'd be clear when viewing the whole file.

**Comment by kinke** [comment]
> Btw, a major obstacle for good CPU scaling of these tests are the `PERMUTE_ARGS`. Each test file is processed in isolation, i.e., the 2^6=64 permutations for `runnable/xtest46.d` are run serially (by `d_do_test`, not managed by make).

@wilzbach: I somehow missed the `test/run.d` tool addition so far. I was thinking about something like that, with the extra addition of parsing the `PERMUTE_ARGS` and distributing/parallelizing the test jobs based on all permutations, probably by using `run_do_test.d` directly (i.e., not invoking its executable - processes come with a significant overhead on Windows). Grouping the output per file would probably be needed too. - Anyway, not sure if I ever get to that, but thx for paving the way!

File: test/Makefile (hunk 3 above)
Comment (kinke) [reply test/Makefile]: I totally missed Iain's #9270; this here is derived from 2.084-based LDC - updated accordingly.

**Comment by wilzbach** [comment]
> Btw, a major obstacle for good CPU scaling of these tests are the `PERMUTE_ARGS`. Each test file is processed in isolation, i.e., the 2^6=64 permutations for `runnable/xtest46.d` are run serially (by `d_do_test`, not managed by make)


Yeah through the argument permutation testing is a bit insane (in terms of being overly cautions and wasting CPU cycles).

> I somehow missed the `test/run.d` tool addition so far. I was thinking about something like that, with the extra addition of parsing the `PERMUTE_ARGS` and distributing/parallelizing the test jobs based on all permutations, probably by using `d_do_test.d` directly [...]

Yeah, `test/run.d` is supposed to replace the Makefile eventually as there are a lot of advantages of doing things in D (not needing to duplicate code for win32 and win64 is another big one).
Another potential improvement would be to replace the ~50 shell scripts with D scripts as we could then build & test DMD without needing any Digital Mars make, GNUmake or even a minGW shell. Though while this is a long-term goal, it will require quite some effort to get there...

**Comment by kinke** [comment]
> not needing to duplicate code for win32 and win64 is another big one

That's not really necessary at the moment either. For LDC, I use GNU make on all platforms, incl. Win32/64, and just had to tweak a few things IIRC. The reason is simple - I haven't known about the existence of a 'Digital Mars make' for years. ;)
